server/channels/app/imaging/preview.go

```diff
@@ -25,26 +25,25 @@ func GeneratePreview(img image.Image, width int) image.Image {
 }
 
 // GenerateThumbnail generates the thumbnail for the given image.
-func GenerateThumbnail(img image.Image, width, height int) image.Image {
+func GenerateThumbnail(img image.Image, targetWidth, targetHeight int) image.Image {
 	thumb := img
-	w := img.Bounds().Dx()
-	h := img.Bounds().Dy()
-	expectedRatio := float64(height) / float64(width)
+	width := img.Bounds().Dx()
+	height := img.Bounds().Dy()
+	expectedRatio := float64(targetHeight) / float64(targetWidth)
 
-	if h > height || w > width {
-		ratio := float64(h) / float64(w)
+	// If both dimensions are over the target size, we scale down until one is not.
```

Comment:
Shouldn't we scale down until both are not? I don't see any recursion/loop so I'd expect that with the current login, if both dimensions are over the target size, we'll end up generating a thumbnail where one dimension is still over target size.

Reply:
I have tried to keep from changing the logic. This comment is based on my understanding of the way the existing algorithm works.

Reply:
@streamer45 while this PR indeed doesn't change the logic, are we fine with scaling down until only one dimension is within limits?

Reply:
The logic is correct, the comment is not. We check the ratio and choose the dimension to scale by based on that.

Reply:
That's the goal yes and the code does that.

Reply:
I still think it doesn't.

Walk through the code with me here.

Line 39:
```
if ratio < expectedRatio {
```
We enter the if clause here if the picture is more wide than the target size we want to scale down to. Like it it's landscape, and the target is portrait. If even despite that, `height >= targetHeight` (line 40), that means `width >= targetWidth` for sure - because not only the height is bigger, the image is more horizontal also.

Then if we start scaling down (line 41), we hit matching heights first. If the other dimension is scaled down proportionately, sure the heights will be the same - but the width will stick out, since we didn't change the aspect ratio. If `targetHeight > height`, though (lines 42-44), then it will scale down to the width - the proportionally bigger coordinate. We know that `targetWidth` is smaller than `width` because of line 37 - since we're in the else clause of line 40, one of the prerequisites for the big if is false. That means the other one is true.

It's a good question whether we want to change that logic entirely, but it's an interface change that will surely be more significant (and potentially breaking) than this small refactor. I'd say it's out of scope.

Reply:
Oh you're absolutely right, sorry! The variable renaming makes the diff really difficult to follow and I got confused about what targetWidth in the original code was actually referring to. So yes, the logic here is only making sure that one of the dimensions matches the target and the other one might still be over. On top of that it's making sure that the image is only ever downscaled and not upscaled if it's already small enough.

I do still think this is intentional and we shouldn't go changing the logic.

Reply:
Actually, no, I can't say anything about what was intended with this originally. Now I'm starting to agree it was probably broken to begin with 🙄

Reply:
@streamer45 can you help us define what's the intention? I'd expect that both dimensions need to be lower than their targets. 

If that's the case, let's create a ticket and a follow up PR to not mix changes. But let's make a decision on the desired behaviour first.

Reply:
@esarafianou The main goal is to resize the input image to fit into the target values while preserving aspect ratio. 

After a security issue we encountered in the past we added more logic to make sure we don't upscale accidentally due to an extreme aspect ratio (see https://github.com/mattermost/mattermost/pull/15534).

But I agree with the above, I'd probably expect this to constrain the dimensions to the input target values in all cases. As things stand, we could be returning very large thumbnails. 

We should file a bug, I'd consider this to be mostly a minor issue as I'd expect clients to resize anyway when rendering and that's likely why this went unnoticed until now but of course it would definitely be more efficient if we don't have to send large thumbnails.

Reply:
I created a ticket here: https://mattermost.atlassian.net/browse/MM-55719. I'm resolving this discussion and approving.